### PR TITLE
kernel/binary_manager : Reload the previous version elf when fail to …

### PR DESCRIPTION
**Pull request by YiHyunJin**
…load the latest version elf

When the latest version can not be loaded, it will exit.
However, we should try to load the previous version.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [f28efa7f765cd5b9293bc1950ac818623cc060dc] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f28efa7f765cd5b9293bc1950ac818623cc060dc] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 1]
```diff
@@ -126,26 +126,38 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
 
-/* Read binary header and update binary table */
-static int binary_manager_load_bininfo(int bin_idx)
+/* Load binary with index in binary table */
+int binary_manager_load_binary(int bin_idx)
 {
 	int ret;
+	pid_t pid;
 	int version;
+	int size;
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: You'd better change size variable type to unsigned int32

**Comment by seinfra** [comment]
Target : [cbfeb4626eee1e6303f9e274fd1c1db70319be7c] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [cbfeb4626eee1e6303f9e274fd1c1db70319be7c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ec26a403eac9fd2f9d287e8940c1634a0928d989] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ec26a403eac9fd2f9d287e8940c1634a0928d989] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [68bb60db7acebdaab209e499e0aff2d1c047b8c8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [68bb60db7acebdaab209e499e0aff2d1c047b8c8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [56e65919d0b6aa771d1adcfd6421f6597219ac28] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [56e65919d0b6aa771d1adcfd6421f6597219ac28] - Code Rule Check Result:
os/kernel/binary_manager/binary_manager_load.c:57: ERROR: [SPC_M_SEP] space required after that close brace '}'

**Comment by seinfra** [comment]
Target : [4839ea90d4064d2cfe63213050e4156ac20a7673] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4839ea90d4064d2cfe63213050e4156ac20a7673] - Code Rule Check Result:
os/kernel/binary_manager/binary_manager_load.c:57: ERROR: [SPC_M_SEP] space required after that close brace '}'

**Comment by seinfra** [comment]
Target : [75578f890d10dd0ab44bf88de93ac1342844b483] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [75578f890d10dd0ab44bf88de93ac1342844b483] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [eeaac5b76d62046db51e9e4634bcce020701ee69] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [eeaac5b76d62046db51e9e4634bcce020701ee69] - Code Rule Check OK.

File: os/kernel/binary_manager/binary_manager_recovery.c [hunk 2]
```diff
@@ -47,7 +47,7 @@ extern bool abort_mode;
  * Private Functions
  ****************************************************************************/
 /* Reboot the board */
-static void binary_manager_board_reset(void)
+void binary_manager_board_reset(void)
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_recovery.c]: please add prototype of this function.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 3]
```diff
@@ -159,19 +173,57 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+		load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;
+	
+		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, load_attr.bin_size, load_attr.offset);
+
+		ret = load_binary(devname, &load_attr);
+		if (ret > 0) {
+			pid = (pid_t)ret;
+			bmvdbg("Load '%s' success! pid = %d\n", devname, pid);
+			break;
+		} else if (ret == -ENOMEM && valid_bin_count == 1) {
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: load_binary will not return -errno.

File: os/kernel/binary_manager/binary_manager.h [hunk 4]
```diff
@@ -57,6 +57,8 @@
 
 #define BINMGR_DEVNAME_FMT         "/dev/mtdblock%d"
 
+#define MAX_WAIT_NUMBER            3                          /* The maximum number of times you can wait to process a delayed free */
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager.h]: how about COUNT instead of NUMBER?

File: os/kernel/binary_manager/binary_manager_load.c [hunk 5]
```diff
@@ -37,6 +37,8 @@
 #include "task/task.h"
 #include "binary_manager.h"
 
+extern volatile sq_queue_t g_delayed_kfree;
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: please add dependency of this variable.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 6]
```diff
@@ -159,19 +173,57 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+		} else if (ret == -ENOMEM && valid_bin_count == 1) {
+			/* Wait until delayed free is processed */
+			while (g_delayed_kfree.head && wait_count < MAX_WAIT_NUMBER) {
+				usleep(500);
+				wait_count++;
+			}
+
+			if (g_delayed_kfree.head == NULL) {
+				continue;
+			}
```
Comment (jeongarmy) [inline os/kernel/binary_manager/binary_manager_load.c]: I think there can be some case that it is available although head is not null.

**Comment by seinfra** [comment]
Target : [927afa4a091ff6655752849d948e1a1cd7b9ec34] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [927afa4a091ff6655752849d948e1a1cd7b9ec34] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d30698768437a8a56c8a1937573b7615edeee1f2] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d30698768437a8a56c8a1937573b7615edeee1f2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [93305b4b8b40f7a272ef565f92bb5dc2821e2976] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [93305b4b8b40f7a272ef565f92bb5dc2821e2976] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_load.c (hunk 5 above)
Comment (YiHyunJin) [reply os/kernel/binary_manager/binary_manager_load.c]: DONE

File: os/kernel/binary_manager/binary_manager_load.c [hunk 7]
```diff
@@ -159,19 +179,63 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+		load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;
+
+		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, load_attr.bin_size, load_attr.offset);
+
+		ret = load_binary(devname, &load_attr);
+		if (ret > 0) {
+			pid = (pid_t)ret;
+			bmvdbg("Load '%s' success! pid = %d\n", devname, pid);
+			break;
+		} else if (ret == ERROR && errno == ENOMEM) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Is `ret == ERROR` necessary?

File: os/kernel/binary_manager/binary_manager_load.c [hunk 8]
```diff
@@ -159,19 +179,63 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+		load_attr.ram_size = header_data[latest_idx].bin_ramsize;
+		load_attr.stack_size = header_data[latest_idx].bin_stacksize;
+		load_attr.priority = header_data[latest_idx].bin_priority;
+		load_attr.offset = CHECKSUM_SIZE + header_data[latest_idx].header_size;
+
+		bmvdbg("BIN[%d] %s %d %d\n", bin_idx, devname, load_attr.bin_size, load_attr.offset);
+
+		ret = load_binary(devname, &load_attr);
+		if (ret > 0) {
+			pid = (pid_t)ret;
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Let's rename `pid` to `pid_loaded` or `loaded_bin_pid` or something else.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 9]
```diff
@@ -159,19 +179,63 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+					usleep(500);
+					wait_count++;
+				}
+
+				if (g_delayed_kfree.head == NULL) {
+					continue;
+				}
+			}
+#endif
+			if (valid_bin_count == 1) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: I don't understand why this conditional necessary.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 10]
```diff
@@ -159,19 +179,63 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+				binary_manager_board_reset();
+			}
+		}
+
+		valid_bin_count--;
+		if (valid_bin_count == 0) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: same as previous

**Comment by seinfra** [comment]
Target : [33109f9bc6eda2930375ddea04ec67ba5814025f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [33109f9bc6eda2930375ddea04ec67ba5814025f] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 11]
```diff
@@ -159,19 +182,62 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+			bin_pid = (pid_t)ret;
+			bmvdbg("Load '%s' success! pid = %d\n", devname, bin_pid);
+			break;
+		} 
+		/* Wait until delayed free is processed */
+#if (defined(CONFIG_BUILD_PROTECTED) || defined(CONFIG_BUILD_KERNEL)) && defined(CONFIG_MM_KERNEL_HEAP)
+		else if (errno == ENOMEM) {
+			if (g_delayed_kfree.head) {
+				wait_count = 0;
+				while (g_delayed_kfree.head && wait_count < MAX_WAIT_COUNT) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Let's leave comments.
```
// wait until delayed free list is empty

File: os/kernel/binary_manager/binary_manager_load.c [hunk 12]
```diff
@@ -159,19 +182,62 @@ static int binary_manager_load_bininfo(int bin_idx)
...
+					usleep(500);
+					wait_count++;
+				}
+
+				if (g_delayed_kfree.head == NULL) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: // If delayed free list is empty, try to load binary again.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 13]
```diff
@@ -159,19 +182,62 @@ static int binary_manager_load_bininfo(int bin_idx)
...
 
 	/* Failed to find valid binary */
-	if (latest_ver < 0) {
+	if (loadable == false) {
 		bmdbg("Failed to find loadable binary %d\n", bin_idx);
 		return ERROR;
 	}
 
+	/* Load binary */
+	while (loadable) {
```
Comment (sunghan-chang) [inline os/kernel/binary_manager/binary_manager_load.c]: Because you already check `loadable` at line 185.
do while is better I think.

**Comment by seinfra** [comment]
Target : [c5b381a90c8d2735a98559a70023fe99afd53ebb] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c5b381a90c8d2735a98559a70023fe99afd53ebb] - Code Rule Check Result:
os/kernel/binary_manager/binary_manager_load.c:234: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [8b2d6340175aeb00caf5a2763d333838af3e817d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8b2d6340175aeb00caf5a2763d333838af3e817d] - Code Rule Check (C++) OK.

**Pull request merged**